### ISSUE TITLE
compiler miscounts lines in s/.../.../e

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2319]
Migrated from [rt.perl.org#3643](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3643) (status was 'resolved')

Searchable as RT3643$

**Comment by p5pRT** [comment]
### From jfriedl@yahoo-inc.com

#### Created by jfriedl@yahoo-inc.com



With versions since 5\.004\, this program \(with line numbers shown\)

&nbsp;       1|    $\_ = "foo";
&nbsp;       2|    s\{
&nbsp;       3|
&nbsp;       4|        f
&nbsp;       5|        o
&nbsp;       6|        o
&nbsp;       7|
&nbsp;       8|    \}\{
&nbsp;       9|
&nbsp;      10|      warn "now";
&nbsp;      11|      warn "now";
&nbsp;      12|      warn "now";
&nbsp;      13|
&nbsp;      14|      "FOO";
&nbsp;      15|    \}gex;
&nbsp;      16|
&nbsp;      17|    warn "now";

prints&#8203;:

&nbsp;    now at /tmp/t3 line 4\.
&nbsp;    now at /tmp/t3 line 5\.
&nbsp;    now at /tmp/t3 line 6\.
&nbsp;    now at /tmp/t3 line 17\.

Perl 5\.003 prints what I would consider to be correct&#8203;:

&nbsp;    now at /tmp/t3 line 10\.
&nbsp;    now at /tmp/t3 line 11\.
&nbsp;    now at /tmp/t3 line 12\.
&nbsp;    now at /tmp/t3 line 17\.

Jeffrey

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by jfriedl at Sat Jul 29 20:09:33 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.15, archname=i686-linux
    uname='linux fummy.dsl.yahoo.com 2.2.16 #6 smp sun jul 23 11:26:16 pdt 2000 i686 unknown '
    config_args='-ds -e -A optimize=-g'
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2 -g', gccversion=pgcc-2.91.66 19990314 (egcs-1.1.2 release)
    cppflags='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /home/jfriedl/lib/perl
    /home/jfriedl/lib/perl/yahoo
    /usr/local/lib/perl5/5.6.0/i686-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/jfriedl
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/local/pgsql/lib:/home/jfriedl/src/rvplayer5.0
    LOGDIR (unset)
    PATH=/home/jfriedl/bin:/home/jfriedl/common/bin:/usr/local/gcc-2.95.2/bin:.:/usr/local/pgsql/bin:/usr/local/bin:/usr/X11R6/bin:/bin:/usr/bin:/usr/sbin:/sbin:/home/jfriedl/src/rvplayer5.0
    PERLLIB=/home/jfriedl/lib/perl:/home/jfriedl/lib/perl/yahoo
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @floatingatoll

I can confirm that this line numbering issue is present in darwin 5\.6\.0 as well \(as 
shipped with OS X\)\.

**Comment by p5pRT** [comment]
### From @floatingatoll

Confirmed present in bleadperl\-threads @&#8203;18374\.

**Comment by p5pRT** [comment]
### From @floatingatoll

Thank you for the bug report\!  It's been stalled\, pending future work on the Perl 
internals\.

This line numbering issue falls into the category of the several instances of Perl 
miscounting line numbers in certain multi\-line circumstances; these issues have been 
assigned to the far\, far future of bugs that will be fixed\, due to lexer issues\.  At this 
time\, I'm tagging the bug as "wontfix"\, low priority; this may change in the future\, once 
the line numbering issues have been addressed globally\.

**Comment by p5pRT** [comment]
@floatingatoll - Status changed from 'open' to 'stalled'

**Comment by p5pRT** [comment]
### From @gannett-ggreer

On Fri Aug 04 18&#8203;:10&#8203;:13 2000\, jfriedl@&#8203;yahoo\-inc\.com wrote&#8203;:
> With versions since 5\.004\, this program \(with line numbers shown\)
> 
>        1|    $\_ = "foo";
>        2|    s\{
>        3|
>        4|        f
>        5|        o
>        6|        o
>        7|
>        8|    \}\{
>        9|
>       10|      warn "now";
>       11|      warn "now";
>       12|      warn "now";
>       13|
>       14|      "FOO";
>       15|    \}gex;
>       16|
>       17|    warn "now";
> 
> prints&#8203;:
> 
>     now at /tmp/t3 line 4\.
>     now at /tmp/t3 line 5\.
>     now at /tmp/t3 line 6\.
>     now at /tmp/t3 line 17\.
> 
> Perl 5\.003 prints what I would consider to be correct&#8203;:
> 
>     now at /tmp/t3 line 10\.
>     now at /tmp/t3 line 11\.
>     now at /tmp/t3 line 12\.
>     now at /tmp/t3 line 17\.

This was an apparently intentional change in perl\-5\.003\_26 \(4fdae800\)
from Chip Salzenbrg titled "Refine 'runaway string' heuristic"\. 'Fixing'
it is as simple as commenting out the line "PL\_multi\_start =
first\_start;" in toke\.c\, although "first\_start" can be removed
completely then\. No tests fail as a result\, although in a way I wish
there were\.

I would prefer "warn" have the exact line\, although there may be more
cases it changes\.

Opinions?

\-\- 
George Greer

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'stalled' to 'open'

**Comment by p5pRT** [comment]
### From @xdg

On Sun\, Jul 25\, 2010 at 7&#8203;:20 PM\, George Greer via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> completely then\. No tests fail as a result\, although in a way I wish
> there were\.
>
> I would prefer "warn" have the exact line\, although there may be more
> cases it changes\.

I agree\, but I'm not sure if I prefer it strong enough to change it
without know what other cases are affected\.

\-\- David

**Comment by p5pRT** [comment]
### From @chipdude



> On Fri Aug 04 18&#8203;:10&#8203;:13 2000\, jfriedl@&#8203;yahoo\-inc\.com wrote&#8203;:
>> With versions since 5\.004\, this program \(with line numbers shown\)
>>
>>        1|    $\_ = "foo";
>>        2|    s\{
>>        3|
>>        4|        f
>>        5|        o
>>        6|        o
>>        7|
>>        8|    \}\{
>>        9|
>>       10|      warn "now";
>>       11|      warn "now";
>>       12|      warn "now";
>>       13|
>>       14|      "FOO";
>>       15|    \}gex;
>>       16|
>>       17|    warn "now";
>>
>> prints&#8203;:
>>
>>     now at /tmp/t3 line 4\.
>>     now at /tmp/t3 line 5\.
>>     now at /tmp/t3 line 6\.
>>     now at /tmp/t3 line 17\.
>>
>> Perl 5\.003 prints what I would consider to be correct&#8203;:
>>
>>     now at /tmp/t3 line 10\.
>>     now at /tmp/t3 line 11\.
>>     now at /tmp/t3 line 12\.
>>     now at /tmp/t3 line 17\.
> This was an apparently intentional change in perl\-5\.003\_26 \(4fdae800\)
> from Chip Salzenbrg titled "Refine 'runaway string' heuristic"\. 'Fixing'
> it is as simple as commenting out the line "PL\_multi\_start =
> first\_start;" in toke\.c\, although "first\_start" can be removed
> completely then\. No tests fail as a result\, although in a way I wish
> there were\.
>
> I would prefer "warn" have the exact line\, although there may be more
> cases it changes\.
>
> Opinions?
>

I just noticed this\.

The purpose of the patch was to fix the "runaway string" warning for
substitutions\.  Try removing the final "\}" in the test code\.  I'll bet
5\.003 says that there's a runaway substituion starting at line 8\, rather
than line 2\.  I had no idea that PL\_multi\_start would affect the line
number diagnostics in s///e\.

I'm inclined to remove my old patch to fix the above problem\.  A more
complete fix would be duplicating PL\_multi\_start to PL\_multi\_startwarn
\(e\.g\.\) and only reset one of them\.

**Comment by p5pRT** [comment]
### From @doy

This still exists in 5\.16\, but 5\.16 also reports the unterminated
substitution at line 8\, rather than line 2\, which Chip indicated should
have been the old behavior\, not the new behavior\, so I don't know what
the correct behavior of that aspect should be\. Also\, 4fdae800 is a large
patch collection from before version control days\, so I don't know which
parts would need to be reverted to test the old behavior\.

\-doy

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in ffdb8b167\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Fixed in ffdb8b167\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
